### PR TITLE
fix: change operation_id from TEXT to INTEGER

### LINKED ISSUE #21893: List operations does not return the full list
### Description

When running 100 actions it's not possible to see the full list, even using limit and offset.

### Juju version

4.0-HEAD

### Cloud

_No response_

### Expected behaviour

```
$ juju operations --offset=50 --limit=50
```

Should return the next 50 items.

### Reproduce / Test

```sh
$ juju bootstrap lxd test
$ juju add-model m
$ juju deploy juju-qa-test
$ for i in {1..100} ; do juju exec -u juju-qa-test/0 --wait 30s "date" ; done
```

### Notes & References

_No response_

### PR DESCRIPTION
**Pull request by tugudd**
The operation_id column in the operations table was TEXT despite cmd expecting an int. With TEXT, ORDER BY sorts lexicographically, which results in numbers like 10 preceding numbers like 2. Cmd also applies numerical sorting on top of the SQL, consequently breaking the pagination.

Fixed by altering the type of operation_id to int in SQL and in code. String IDs are parsed to uint64 in the service layer; domain types stay string for facade compatibility.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~Integration tests~
- [ ] ~doc.go added or updated in changed packages~

## QA steps

```sh
 go test -v -run 'TestQuerySuite/TestGetOperationByID' ./domain/operation/service/...
 ```

```sh
juju bootstrap lxd test
juju add-model m
juju deploy juju-qa-test
for i in {1..20} ; do juju exec -u juju-qa-test/0 --wait 30s "date" ; done
juju operations --offset 1
juju operations --offset 2
```

You might see the operation with id 2 at the top with offset 1. With offset 2, this operation should leave. In older versions, the operation with id 10 was left out.

## Documentation changes

N/A

## Links

**Issue:** Fixes #21893.

**Jira card:** [JUJU-9279](https://warthogs.atlassian.net/browse/JUJU-9279)

[JUJU-9279]: https://warthogs.atlassian.net/browse/JUJU-9279?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: domain/schema/model/sql/0062-operation.PATCH.sql [hunk 1]
```diff
@@ -0,0 +1,27 @@
+-- Change operation_id from TEXT to INTEGER. The values are generated by a
+-- monotonic uint64 sequence (domain/sequence/state/sequence.go:24) and were
+-- previously stringified before storage. Storing as INTEGER matches the true
+-- type and fixes lexicographic-vs-numeric ordering of ORDER BY operation_id
+-- (domain/operation/state/query.go:307) once IDs reach double digits.
+CREATE TABLE operation_new (
```
Comment (nvinuesa) [inline domain/schema/model/sql/0062-operation.PATCH.sql:49]: Why not `ALTER TABLE operation`? we just need to change the type of `operation_id` right?

**Review by nvinuesa — COMMENTED** [review]
First pass, looking good. I do have doubts about the SQL patch file.

File: domain/schema/model/sql/0062-operation.PATCH.sql (hunk 1 above)
Comment (tugudd) [reply domain/schema/model/sql/0062-operation.PATCH.sql:49]: As far as I know, you cannot modify the type in SQLite.

Comment (tugudd) [reply domain/schema/model/sql/0062-operation.PATCH.sql:49]: Plus, I have checked other SQL files, and this is how it was done.

Comment (nvinuesa) [reply domain/schema/model/sql/0062-operation.PATCH.sql:49]: Ah yes right, we need to do the cast and then rename.

File: domain/export/version.go [hunk 2]
```diff
@@ -18,7 +18,8 @@ import (
 var exportVersionStrings = []string{
 	"4.0.4",
 	"4.0.6",
-	"4.0.12", // version currently in dev.
+	"4.0.12",
+	"4.0.13", // version currently in dev.
```
Comment (nvinuesa) [inline domain/export/version.go:22]: I'll add some infra for this, but we will only need one version here (we don't allow 4.0->4.0 migrations), which is the last one that hasn't any schema changes. Nothing to do for you though.

Comment (tugudd) [reply domain/export/version.go:22]: Got it

File: domain/schema/model/sql/0062-operation.PATCH.sql [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    CAST(operation_id AS INTEGER) AS operation_id,
+    summary,
+    enqueued_at,
+    started_at,
+    completed_at,
+    parallel,
+    execution_group
+FROM operation;
+
+DROP TABLE operation;
```
Comment (gfouillet) [inline domain/schema/model/sql/0062-operation.PATCH.sql:307]: **issue:** `DROP TABLE operation` will fail on any existing model that has operations with linked tasks, actions, or parameters. Three child tables (`operation_action`, `operation_task`, `operation_parameter`) have `FOREIGN KEY (operation_uuid) REFERENCES operation (uuid)` with `NO ACTION`. Since `PRAGMA foreign_keys = ON` is set before schema migration runs (`internal/database/bootstrap.go:136`), SQLite performs an implicit `DELETE FROM operation` before the DROP, which triggers FK constraint violations.

The established precedent is `0056-blockdevice-partial.PATCH.sql`, which explicitly documents this issue and rebuilds all child tables with FKs pointing to the `_new` table before dropping the parent. Schema tests pass only because they run on empty databases.

File: domain/operation/service/query.go [hunk 4]
```diff
@@ -72,7 +73,12 @@ func (s *Service) GetOperationByID(ctx context.Context, operationID string) (ope
 	ctx, span := trace.Start(ctx, trace.NameFromFunc())
 	defer span.End()
 
-	res, err := s.st.GetOperationByID(ctx, operationID)
+	id, err := strconv.ParseUint(operationID, 10, 64)
+	if err != nil {
+		return operation.OperationInfo{}, errors.Errorf("invalid operation ID %q: %w", operationID, err)
```
Comment (gfouillet) [inline domain/operation/service/query.go]: **todo:** Missing `.Add(coreerrors.NotValid)` on this error. The same file uses it at line 138 for the same pattern (`errors.Errorf(...).Add(coreerrors.NotValid)`). Without it, callers can't use `errors.Is(err, coreerrors.NotValid)` to distinguish a malformed ID from not-found or internal errors.

File: domain/operation/modelmigration/import.go [hunk 5]
```diff
@@ -160,8 +161,13 @@ func (i *importOperation) buildImportArgs(ctx context.Context, model description
 		tasks, _ := tasksByOp[op.Id()]
 		delete(tasksByOp, op.Id())
 
+		opID, err := strconv.ParseUint(op.Id(), 10, 64)
```
Comment (gfouillet) [inline domain/operation/modelmigration/import.go:164]: **todo:** No test exercises this `ParseUint` error path. All migration tests use valid numeric IDs ("1", "2"). Add a test with a non-numeric operation ID (e.g. `"op-abc"`) to verify the error is returned and `importService.InsertMigratingOperations` is never called.

File: domain/schema/model/sql/0062-operation.PATCH.sql [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    execution_group TEXT
+);
+
+INSERT INTO operation_new (
+    uuid, operation_id, summary, enqueued_at, started_at,
+    completed_at, parallel, execution_group
+)
+SELECT
+    uuid,
+    CAST(operation_id AS INTEGER) AS operation_id,
```
Comment (gfouillet) [inline domain/schema/model/sql/0062-operation.PATCH.sql:66]: **todo:** No test verifies the CAST behavior. SQLite's `CAST('abc' AS INTEGER)` silently returns 0 — if two non-numeric values existed, they'd both CAST to 0 and the unique index creation would fail with an opaque error. Add a schema test that inserts a TEXT `operation_id` and verifies the post-migration type is `integer` with the correct value.

File: domain/operation/state/query.go [hunk 7]
```diff
@@ -222,7 +223,7 @@ func encodeOperationInfo(
 	taskLogs map[string][]taskLogEntryByOperation,
 ) (operation.OperationInfo, error) {
 	var opInfo operation.OperationInfo
-	opInfo.OperationID = op.OperationID
+	opInfo.OperationID = strconv.FormatUint(op.OperationID, 10)
```
Comment (gfouillet) [inline domain/operation/state/query.go:226]: **todo:** No test verifies that `GetOperations` returns results in numeric (not lexicographic) order — the PR's stated purpose. The test helper `sequenceGenerator()` starts at 0, so IDs sort identically under TEXT and INTEGER. Add a test with >=10 operations asserting `operation_id` order is 1, 2, ..., 10, 11, 12 (not 1, 10, 11, ..., 2, ...).

File: domain/operation/internal/types.go [hunk 8]
```diff
@@ -41,7 +41,7 @@ type CompletedTask struct {
 type ImportOperationsArgs []ImportOperationArg
 
 type ImportOperationArg struct {
-	ID        string
+	ID        uint64
```
Comment (gfouillet) [inline domain/operation/internal/types.go:44]: **nitpick:** The comment at line 38 still says "numeric/string ID" but `ImportOperationArg.ID` was changed from `string` to `uint64` in this PR. The "string" qualifier is now misleading.

File: domain/schema/model/sql/0062-operation.PATCH.sql (hunk 6 above)
Comment (gfouillet) [inline domain/schema/model/sql/0062-operation.PATCH.sql:66]: **suggestion:** `CAST(operation_id AS INTEGER)` silently converts non-numeric strings to 0. In practice operation_ids are always numeric, but consider adding a pre-migration validation query (e.g. `SELECT COUNT(*) FROM operation WHERE operation_id != CAST(operation_id AS INTEGER)`) to provide a clearer error if anomalous data exists.

File: domain/operation/service/query_test.go [hunk 9]
```diff
@@ -183,14 +183,33 @@ func (s *querySuite) TestGetOperationByIDError(c *tc.C) {
 	expectedErr := errors.New("not found")
 	s.state.EXPECT().GetOperationByID(gomock.Any(), gomock.Any()).Return(operation.OperationInfo{}, expectedErr)
 
-	_, err := s.service(c).GetOperationByID(c.Context(), "unknown")
+	_, err := s.service(c).GetOperationByID(c.Context(), "999")
 	c.Assert(err, tc.ErrorMatches, "not found")
 }
 
+func (s *querySuite) TestGetOperationByIDInvalidFormat(c *tc.C) {
```
Comment (gfouillet) [inline domain/operation/service/query_test.go:190]: **suggestion:** `TestGetOperationByIDInvalidFormat` covers non-numeric input, but no test exercises `ParseUint` overflow (e.g. `"99999999999999999999999"`) or negative values (`"-1"`). Consider adding these to cover different error paths.

File: domain/export/state/model/export_test.go [hunk 10]
```diff
@@ -13,15 +13,15 @@ import (
...
 }
 
-func TestExportStateSuiteV4_0_12(t *testing.T) {
-	tc.Run(t, &exportStateSuiteV4_0_12{})
+func TestExportStateSuiteV4_0_13(t *testing.T) {
+	tc.Run(t, &exportStateSuiteV4_0_13{})
 }
 
-func (s *exportStateSuiteV4_0_12) TestExportRuns(c *tc.C) {
+func (s *exportStateSuiteV4_0_13) TestExportRuns(c *tc.C) {
```
Comment (gfouillet) [inline domain/export/state/model/export_test.go:24]: **suggestion:** `TestExportRuns` only tests an empty database. Add a test that inserts an operation row with a known `operation_id` and verifies the export reads it correctly as `int64` from the new INTEGER column.

File: domain/operation/state/start.go [hunk 11]
```diff
@@ -129,7 +129,7 @@ func (st *State) AddActionOperation(ctx context.Context,
 		// Insert the operation first.
 		err = st.insertOperation(ctx, tx, insertOperation{
 			UUID:           operationUUID.String(),
-			OperationID:    strconv.FormatUint(operationID, 10),
+			OperationID:    operationID,
```
Comment (gfouillet) [inline domain/operation/state/start.go:132]: **nitpick:** `start_test.go` (lines 671, 693) still passes `result.OperationID` (a `string`) for `WHERE operation_id = ?` on the now-INTEGER column. Works via SQLite type affinity but is a code smell. Consider parsing to `uint64` before querying in test helpers.

File: domain/schema/model/sql/0062-operation.PATCH.sql (hunk 3 above)
Comment (tugudd) [reply domain/schema/model/sql/0062-operation.PATCH.sql:307]: Yeah, this one. I forgot that FKs are used in the code.

**Comment by tugudd** [comment]
Obsolete
